File: TPS_Exercicios/tp01/TP01Q12 - RECURSIVO - Palíndromo em C/palindromoRecursivo.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <locale.h>

int ContaChars(char string[]) {
    int i = 0;
    while (string[i] != '\0') {
        i++;
    }
    return i;
}

// Verifica se a string analisada é a "FIM"
bool isFim(char entrada[]) {
    return (entrada[0] == 'F' && entrada[1] == 'I' && entrada[2] == 'M' && entrada[3] == '\0');
}

// Verifica se e polindromo de forma recursiva
void classificarPalindromo(char palavra[], int esq, int dir, bool *resultado) {
    if (esq < dir) {
        if (palavra[esq] != palavra[dir]) {
            *resultado = false; 
        } else {
            classificarPalindromo(palavra, esq + 1, dir - 1, resultado);
        }
    }
}

int main() {    
    setlocale(LC_ALL, "pt_BR.UTF-8");  // Linux

    char palavra[1000];

    scanf(" %[^\n]", palavra);

    while (!isFim(palavra)) {

        for (int i = 0; palavra[i] != '\0'; i++) {
            if (!((palavra[i] <= 'Z' && palavra[i] >= 'A') || (palavra[i] <= 'z' && palavra[i] >= 'a') || (palavra[i] <= '9' && palavra[i] >= '0'))) {
                palavra[i] = 'A';
            }
        }

        bool resultado = true;
        classificarPalindromo(palavra, 0, ContaChars(palavra) - 1, &resultado);

        if (resultado) {
            printf("SIM\n");
        } else {
            printf("NAO\n");
        }

        scanf(" %[^\n]", palavra);
    }

    return 0;
}

File: TPS_Exercicios/tp01/TP01Q11 - RECURSIVO - Palíndromo em Java/palindromoRecursivo.java
import java.util.*;

class palindromoRecursivo {

    public static boolean isFim(String entrada) {
        return (entrada.length() == 3 && entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M');
    }

    public static String verificaPolindromo(String polindromo, int esq, int dir) {
        String resposta = "SIM";

        if (esq < dir) {
            if (polindromo.charAt(esq) != polindromo.charAt(dir)) {
                resposta = "NAO";
            } else {
                resposta = verificaPolindromo(polindromo, esq + 1, dir - 1);
            }
        }

        return resposta;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String entrada[] = new String[1000];
        int i = 0;

        do {
            entrada[i] = sc.nextLine();
        } while (!isFim(entrada[i++]));

        for (int j = 0; j < i - 1; j++) {
            System.out.println(verificaPolindromo(entrada[j], 0, entrada[j].length() - 1));
        }
    }
}
